src/gtc/passes/oir_optimizations/horizontal_execution_merging.py

```diff
@@ -100,3 +79,140 @@ def visit_VerticalLoopSection(
                 expected="the number of horizontal executions is equal or smaller than before"
             )
         return result
+
+
+class OnTheFlyMerging(NodeTranslator):
+    """Merges consecutive horizontal executions inside parallel vertical loops by introducing redundant computations."""
+
+    def visit_CartesianOffset(
+        self,
+        node: common.CartesianOffset,
+        *,
+        shift: Optional[Tuple[int, int, int]] = None,
+        **kwargs: Any,
+    ) -> common.CartesianOffset:
+        if shift:
+            di, dj, dk = shift
+            return common.CartesianOffset(i=node.i + di, j=node.j + dj, k=node.k + dk)
+        return self.generic_visit(node, **kwargs)
+
+    def visit_FieldAccess(
+        self,
+        node: oir.FieldAccess,
+        *,
+        offset_symbol_map: Dict[Tuple[str, Tuple[int, int, int]], str] = None,
+        **kwargs: Any,
+    ) -> Union[oir.FieldAccess, oir.ScalarAccess]:
+        if offset_symbol_map:
+            offset = self.visit(node.offset, **kwargs)
+            key = node.name, (offset.i, offset.j, offset.k)
+            if key in offset_symbol_map:
+                return oir.ScalarAccess(name=offset_symbol_map[key], dtype=node.dtype)
+        return self.generic_visit(node, **kwargs)
+
+    def _merge(
+        self,
+        horizontal_executions: List[oir.HorizontalExecution],
+        symtable: Dict[str, Any],
+        tmps_to_remove: Set[str],
+        new_symbol_name: Callable[[str], str],
+    ) -> List[oir.HorizontalExecution]:
+        """Recursively merge horizontal executions.
+
+        Uses the following algorithm:
+        1. Get output fields of the first horizontal execution.
+        2. Check in which following h. execs. the outputs are read.
+        3. Duplicate the body of the first h. exec. for each read access (with corresponding offset) and prepend it to the depending h. execs.
+        4. Recurse on the resulting h. execs.
+        """
```

Comment:
describe limitations:
- copies whole h exec, even if only a few statements would be relevant
- heuristic would be needed for size and for heavy functions (e.g. `exp`)

Reply:
Added comment and configurable merging limitations (body size and expensive functions).